core/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java

```diff
@@ -0,0 +1,108 @@
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.Objects;
+
+/**
+ * {@code DeprecationRestHandler} provides a proxy for any existing {@link RestHandler} so that usage of the handler can be
+ * logged using the {@link DeprecationLogger}.
+ * <p>
+ * In addition, all responses are handled with the {@link DeprecationRestChannel} so that responses can be instrumented with
+ * deprecation warnings.
+ */
+public class DeprecationRestHandler implements RestHandler {
+    /**
+     * The proxied {@link RestHandler} that gets executed.
+     */
+    private final RestHandler handler;
+    /**
+     * The message to display upon using the {@link #handler} in ES logs.
+     */
+    private final String warning;
+    /**
+     * The deprecation logger used to warn within Elasticsearch for any usage.
+     */
+    private final DeprecationLogger deprecationLogger;
+
+    /**
+     * Create a {@link DeprecationRestHandler} that encapsulates the {@code handler} with its {@code Class} to log
+     * deprecation {@code warning}.
+     *
+     * @param handler The rest handler to deprecate (it's possible that the handler is reused with a different name!)
+     * @param warning The message to warn users with when they use the {@code handler}
+     * @param clazz The type of rest handler
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestHandler(RestHandler handler, String warning, Class<? extends RestHandler> clazz) {
+        this(handler, warning, Loggers.getLogger(clazz));
+    }
+
+    /**
+     * Create a {@link DeprecationRestHandler} that encapsulates the {@code handler} using its {@code logger} to log
+     * deprecation {@code warning}.
+     *
+     * @param handler The rest handler to deprecate (it's possible that the handler is reused with a different name!)
+     * @param warning The message to warn users with when they use the {@code handler}
+     * @param logger The logger to use for deprecation
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestHandler(RestHandler handler, String warning, ESLogger logger) {
+        this(handler, warning, new DeprecationLogger(logger));
+    }
+
+    /**
+     * Create a {@link DeprecationRestHandler} that encapsulates the {@code handler} using the {@code deprecationLogger} to log
+     * deprecation {@code warning}.
+     *
+     * @param handler The rest handler to deprecate (it's possible that the handler is reused with a different name!)
+     * @param warning The message to warn users with when they use the {@code handler}
+     * @param deprecationLogger The deprecation logger
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestHandler(RestHandler handler, String warning, DeprecationLogger deprecationLogger) {
+        // it had better not be empty or you're deprecating it wrong
+        assert Strings.hasText(warning);
```

Comment:
I thought about it, but this is just logging. It's not until we get to the channel where it changes purpose and sends a header.

Though I do prefer that it fails fast instead of lazily later.